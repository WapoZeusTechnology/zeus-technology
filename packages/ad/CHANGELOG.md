# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.0.0 (2019-08-30)

**Note:** Version bump only for package @zeus-platform/ad

## [**1.0.8**] - 2019-09-10

### Changed

* Newly licensed under the MIT license
* New public repository

## [**1.0.9**] - 2019-09-23

### Added

* Support for navigation in single-page-applications so that by default the ad changes whenever the route changes. This only supports `react-router-dom` and compatible variants. Use the `ZeusAdWithRouter` component to take advantage of this.

## [**1.0.10**] - 2019-10-04

### Fixed

* Dependency issue with `react-router-dom` is now repaired.

### Changed

* Re-factored components to be more module and testable.

### Added

* Travis-CI support!

## [**1.0.10**] - 2019-10-08

### Added

* Support for disabling the `react-router-dom`-based ad slot refreshing.
* Support for screen-based granularity in key-value pairs.

### Fixed

* Issue where ad re-rendering was counting as a refresh.
