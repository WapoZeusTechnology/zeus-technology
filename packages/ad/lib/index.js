import React from "react";
/**
 * This component introduces a `zeus-ad` element to present an ad inventory slot.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {string} props.slotId The ID of the inventory slot.
 */
export const ZeusAd = props => {
  return (
    <div style={{ border: "10px solid red" }}>
      <zeus-ad id={props.slotId} {...props} />
    </div>
  );
};
