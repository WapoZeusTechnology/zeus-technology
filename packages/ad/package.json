{
  "name": "@zeus-technology/ad",
  "private": false,
  "version": "1.0.3",
  "description": "The Ad slot component for Zeus",
  "author": "Mike Stemle <mike.stemle@washpost.com>",
  "homepage": "https://github.com/WPMedia/zeus-react-public#readme",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "publishConfig": {
    "directory": "dist",
    "access": "public"
  },
  "directories": {
    "lib": "dist",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/WPMedia/zeus-react-public.git"
  },
  "scripts": {
    "build": "BABEL_ENV=build babel . --out-dir dist --source-maps --delete-dir-on-start --config-file ./babel.config.js --no-comments --copy-files",
    "clean": "rm -rf dist/*",
    "test": "jest"
  },
  "bugs": {
    "url": "https://github.com/WPMedia/zeus-react-public/issues"
  },
  "dependencies": {
    "prop-types": "^15.7.2",
    "react": "^16.1.0"
  },
  "gitHead": "4969a530fd17a8194300272698f9c6083628c951"
}
