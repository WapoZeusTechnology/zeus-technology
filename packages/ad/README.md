# `@zeus-technology/ad`

[![Build Status](https://travis-ci.org/WapoZeusTechnology/zeus-technology.svg?branch=master)](https://travis-ci.org/WapoZeusTechnology/zeus-technology)

The purpose of the `ZeusAd` component is to embed a Zeus-managed advertisement into a React app.

**NOTE**: Any parts of this module which mention a "router," those have been tested with `react-router-dom`, and are peer dependencies. If you are not using the `ZeusAdWithRouter` component, you need not install this dependency.

If you need additional routers supported, please submit an issue.

## Usage

### `ZeusAd`

```js
import React, { Fragment } from ‘react’;
import { ZeusAd } from '@zeus-technology/ad';

const App = (props) => (
	<Fragment>
		<ZeusAd slotId='ad-slot-name' style={{display:visible?'':'none'}} />
	</Fragment>
)
```

### `ZeusAdWithRouter`

If you're using `react-router-dom` and you would like to have your ads
refresh when your route changes, try this. With this module, your Zeus
ad slot can re-render (not refresh) with every change of the route.

The default behavior is that every change of the route will cause the
ad slot to re-render. Using the _Boolean control_, you can bind a state
variable to the component to prevent it from refreshing in certain state
conditions. Using the _function control_, you are able to have a function
which gives a finer degree of control over this behavior.

#### Boolean control

This control is achieved using the `changeOnNavigate` property.

```js
import React, { Fragment } from ‘react’;
import { ZeusAdWithRouter } from '@zeus-technology/ad';

const App = (props) => (
	<Router>
		<ZeusAdWithRouter changeOnNavigate={state.variable} slotId='ad-slot-name' />
	</Router>
)
```

The optional `changeOnNavigate` prop defaults to `true`.

#### Function control

This control is implemented by way of a function that you supply as the
`shouldChangeForRoute` prop. This function takes two arguments: the current location,
and the former location, as supplied by `react-router-dom`.

```js
import React, { Fragment } from ‘react’;
import { ZeusAdWithRouter } from '@zeus-technology/ad';

const App = (props) => (
	<Router>
		<ZeusAdWithRouter
			slotId='ad-slot-name'
			shouldChangeForRoute={(current,previous) => !current.search.match(/refresh=skip/)} />
	</Router>
)
```

In this example we're telling `ZeusAdWithRouter` only to change for routes which don't have
search queries matching `refresh=skip`.

### `ZeusRouteResponder`

This component allows you to hook in to your route to communicate key-value pairs to Zeus.
Just like `ZeusAdWithRouter`, you can use the `shouldChangeForRoute` prop to pass a function
which allows you to control whether or not this event shall fire.

This is useful for cases where you wish for key-value pairs to be supplemented with your
current "page" in a single-page application. By using this component within a route, you
can notify Zeus that you have new key-value pairs.

```js
import React, { Fragment } from ‘react’;
import { ZeusRouteResponder } from '@zeus-technology/ad';

const App = (props) => (
	<Router>
		<ZeusAdWithRouter
			shouldChangeForRoute={(current,previous) => !current.search.match(/refresh=skip/)}
			keyValuePairs={state.pageKvps} />
	</Router>
)
```

You may also use this component without the optional `keyValuePairs` property if you
already have all of your key-value pairs configured within your hosted Zeus library.

## See Also

- [Documentation for react-router-dom](https://reacttraining.com/react-router/web/guides/quick-start)
