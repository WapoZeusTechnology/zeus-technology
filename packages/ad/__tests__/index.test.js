"use strict";

import React from "react";
import { ZeusAd } from "@zeus-platform/ad";
import renderer from "react-test-renderer";

describe("@zeus-platform/ad", () => {
  it("Should render and match snapshot", () => {
    const sample = renderer.create(<ZeusAd slotId="sample0" />);
    expect(sample.toJSON()).toMatchSnapshot();
  });
});
