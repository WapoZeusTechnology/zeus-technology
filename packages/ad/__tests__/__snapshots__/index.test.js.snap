// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@zeus-platform/ad Should render and match snapshot 1`] = `
<div
  style={
    Object {
      "border": "10px solid red",
    }
  }
>
  <zeus-ad
    id="sample0"
    slotId="sample0"
  />
</div>
`;
